src/backend_model.cc

```diff
@@ -114,42 +114,58 @@ TritonModel::Create(
 
   std::string backend_libdir;
   std::string backend_libpath;
-  for (const auto& path : search_paths) {
-    const auto full_path = JoinPath({path, backend_libname});
+  RETURN_IF_ERROR(LocateBackendLibrary(
+      search_paths, backend_libname, backend_libdir, backend_libpath));
+
+  std::string py_backend_path;
+  if (backend_libpath.empty()) {
+    // Check if specialized backend is a python-based backend
+    // and model.py is present in the specialized backend's directory
+    const std::string specialized_backend_path =
+        JoinPath({backend_dir, specialized_backend_name, "model.py"});
     bool exists = false;
-    RETURN_IF_ERROR(FileExists(full_path, &exists));
+    RETURN_IF_ERROR(FileExists(specialized_backend_path, &exists));
+
     if (exists) {
-      backend_libdir = path;
-      backend_libpath = full_path;
-      break;
+      py_backend_path = JoinPath({backend_dir, specialized_backend_name});
+      RETURN_IF_ERROR(server->BackendManager()->StoreCustomBackend(
+          backend_name, specialized_backend_path));
+      // Locate python backend's shared library
+      specialized_backend_name = "python";
+      backend_name = "python";
+      RETURN_IF_ERROR(BackendConfigurationBackendLibraryName(
+          specialized_backend_name, &backend_libname));
+      search_paths.emplace_back(
+          JoinPath({backend_dir, specialized_backend_name}));
+      RETURN_IF_ERROR(LocateBackendLibrary(
+          search_paths, backend_libname, backend_libdir, backend_libpath));
+    } else {
+      return Status(
+          Status::Code::INVALID_ARG,
+          "unable to find '" + backend_libname + "' or '" +
+              specialized_backend_name + "/model.py' for model '" +
+              model_config.name() + "', searched: " + version_path + ", " +
+              model_path + ", " + global_path + ", " +
+              specialized_backend_path);
     }
   }
 
-  if (backend_libpath.empty()) {
-    return Status(
-        Status::Code::INVALID_ARG, "unable to find '" + backend_libname +
-                                       "' for model '" + model_config.name() +
-                                       "', searched: " + version_path + ", " +
-                                       model_path + ", " + global_path);
-  }
-
   // Resolve the global backend configuration with the specific backend
   // configuration
   triton::common::BackendCmdlineConfig config;
-  RETURN_IF_ERROR(ResolveBackendConfigs(
-      backend_cmdline_config_map, model_config.backend(), config));
+  RETURN_IF_ERROR(
+      ResolveBackendConfigs(backend_cmdline_config_map, backend_name, config));
 
   RETURN_IF_ERROR(SetBackendConfigDefaults(config));
 
   std::shared_ptr<TritonBackend> backend;
   RETURN_IF_ERROR(server->BackendManager()->CreateBackend(
-      model_config.backend(), backend_libdir, backend_libpath, config,
-      &backend));
+      backend_name, backend_libdir, backend_libpath, config, &backend));
```

Comment:
Question: could we modify `CreateBackend`, and the `TritonBackend` class with additional flag to indicate a `python_backend`?

Wondering if instead of adding path to the model or having a new custom map - we could somehow augment the Backend Creation  / Backend class to store the additional information ....

Would that allow for the artifact directory to return the python folder (if the backend class had a flag) and avoid the need for a new api?

Reply:
This is essentially what I was getting at in my other comment. I think the first iteration of this can be done without introducing any new APIs with a little refactoring.

Reply:
I would like to point out that, when backend is being created, it checks if the same backend was already loaded in `backend_map`, if it was, then model is associated with this already created backend. Thus, if we want to make sure that Triton handles `python` model + `vllm` model, for both of these instances we should have a separate copy of a backend then. One standard, one with specified location to `model.py`.  If this is the route, then probably on the `python backend` side I can add a special handling when `TRITONBACKEND_BackendName` returns non-`python` backend name. I can explore that route. A quick disadwantage I see, is multeple copies of essentially the same backend

Reply:
Basically, some special logic here to see if a backend is a true `python` backend or not: https://github.com/triton-inference-server/python_backend/blob/b136bf3fad77d5d467a6db5bd739be16eb9d9400/src/python_be.cc#L1722

Reply:
cc @tanmayv25 this is basically goes back to our discussion on where to keep path to `model.py`.

Reply:
Just an update. There will be no special logic, since backend-model is 1 to 1 relationship. Thus I would need to copy python backend's `dlopen / dlsym handles` into custom backend, which I referred to earlier as `multiple copies of the same backend)